### ISSUE TITLE
How to use metrics for multi-class binary mask target and multi-class multi-channel output?

### ISSUE THREAD

**Issue by sarmientoj24** [qubvel/segmentation_models.pytorch#541]
I saw in the documentation that the metrics for multilabel prediction require `(batch, num_class, height, width)`.
But then, I have a multi-class mask of one channel as target where each pixel are labeled by the class. 

How do I use this for that scenario?

Also, this seems to be computing it per batch. How do I do it per epoch? 
 
```
import segmentation_models_pytorch as smp

# lets assume we have multilabel prediction for 3 classes
output = torch.rand([10, 3, 256, 256])
target = torch.rand([10, 3, 256, 256]).round().long()

# first compute statistics for true positives, false positives, false negative and
# true negative "pixels"
tp, fp, fn, tn = smp.metrics.get_stats(output, target, mode='multilabel', threshold=0.5)

# then compute metrics with required reduction (see metric docs)
iou_score = smp.metrics.iou_score(tp, fp, fn, tn, reduction="micro")
f1_score = smp.metrics.f1_score(tp, fp, fn, tn, reduction="micro")
f2_score = smp.metrics.fbeta_score(tp, fp, fn, tn, beta=2, reduction="micro")
accuracy = smp.metrics.accuracy(tp, fp, fn, tn, reduction="macro")
recall = smp.metrics.recall(tp, fp, fn, tn, reduction="micro-imagewise")
```

**Comment by qubvel** [comment]
1. For multiclass it almost the same

```python
import segmentation_models_pytorch as smp

target = torch.rand([10, 256, 256]).round().long()

output = torch.rand([10, 3, 256, 256])  
output = torch.argmax(output, dim=1)  # convert with argmax to shape of target

# first compute statistics for true positives, false positives, false negative and
# true negative "pixels"
tp, fp, fn, tn = smp.metrics.get_stats(output, target, mode='multiclass')

# compute metric
iou_score = smp.metrics.iou_score(tp, fp, fn, tn, reduction="micro")
```

2. To compute it per epoch see the example: 
Training model for pets binary segmentation with Pytorch-Lightning [notebook](https://github.com/qubvel/segmentation_models.pytorch/blob/master/examples/binary_segmentation_intro.ipynb) and [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/googlecolab/colabtools/blob/master/examples/binary_segmentation_intro.ipynb)

**Comment by sarmientoj24** [comment]
Thank you for the quick response. I tried this but it seems to be giving me a score of 1 on all metrics such as `iou_score`, `f1`, 'accuracy`.

**Comment by qubvel** [comment]
Could you upload `npy` files with example of your target and output tensors? Ans a code sample?

**Comment by sarmientoj24** [comment]
This is what my reconstructed target and output looks like 
![image](https://user-images.githubusercontent.com/8830319/150517802-c7b13558-b125-46a2-bdbb-dad071a710e6.png)

Here is the validation code for pytorch lightning
```
 def validation_step(self, batch, batch_idx):
        img, mask = batch
        img = img.float()
        mask = mask.long()
        out = self(img)
        loss_val = F.cross_entropy(out, mask, ignore_index=250)
        
        # self.log("val_loss", loss_val, on_step=True, on_epoch=True)
        
        # Metrics
        output_mask = torch.nn.functional.softmax(
            out.to(torch.float32), dim=1
        )
        output_mask = torch.argmax(output_mask, dim=1)
        
        tp, fp, fn, tn = smp.metrics.get_stats(
            output_mask.long(), mask.long(), 
            mode="multiclass", num_classes=9
        )
        
        # Logging mechanism (You can ignore this.)
        num_images = min(len(img), 4)

        if batch_idx == 0:
            orig_images = []
            recon_images = []

            for i in range(0, num_images):
                original_img = img[i].permute(1, 2, 0).detach().cpu().numpy()

                binary_mask = mask[i].detach().cpu().numpy()
                binary_mask = convert_binary_mask(binary_mask, colors=self.label_colors)

                reconstructed_mask = out[i].detach()
                reconstructed_mask = torch.nn.functional.softmax(
                    reconstructed_mask.to(torch.float32), dim=1
                )
                reconstructed_mask = torch.argmax(reconstructed_mask, dim=0)
                reconstructed_mask = reconstructed_mask.cpu().numpy()
                reconstructed_mask = add_plus_one_except(reconstructed_mask)
                reconstructed_mask = convert_binary_mask(
                    reconstructed_mask, colors=self.label_colors
                )

                orig_images.append(original_img)
                recon_images.append(
                    np.concatenate((binary_mask, reconstructed_mask), axis=1)
                )

            self.logger.log_image(key="images", images=orig_images)
            self.logger.log_image(key="mask_vs_recon", images=recon_images)

        return {
            "val_loss": loss_val,
            "tp": tp,
            "fp": fp,
            "fn": fn,
            "tn": tn,
        }

    def validation_epoch_end(self, outputs):
        loss_val = torch.stack([x["val_loss"] for x in outputs]).mean()
        
        tp = torch.cat([x["tp"] for x in outputs])
        fp = torch.cat([x["fp"] for x in outputs])
        fn = torch.cat([x["fn"] for x in outputs])
        tn = torch.cat([x["tn"] for x in outputs])

        iou_score = smp.metrics.iou_score(tp, fp, fn, tn, reduction="micro")
        f1_score = smp.metrics.f1_score(tp, fp, fn, tn, reduction="micro")
        f2_score = smp.metrics.fbeta_score(tp, fp, fn, tn, beta=2, reduction="micro")
        accuracy = smp.metrics.accuracy(tp, fp, fn, tn, reduction="macro")
        recall = smp.metrics.recall(tp, fp, fn, tn, reduction="micro-imagewise")
        
        log_dict = {"val_loss": loss_val}
        
        self.log("val_loss", loss_val, on_epoch=True)
        self.log("f1_score", f1_score, on_epoch=True)
        self.log("f2_score", f2_score, on_epoch=True)
        self.log("accuracy", accuracy, on_epoch=True)
        self.log("recall", recall, on_epoch=True)
        
        return {"log": log_dict, "val_loss": log_dict["val_loss"]}
```

**Comment by sarmientoj24** [comment]
Here are the files
- [masks](https://drive.google.com/file/d/1bcjTOiia-h6jd702QjCi_YnUKJKR6Eef/view?usp=sharing)
- [output](https://drive.google.com/file/d/1DXMS8KwfqHdj4Lo-9LVME7rPWvEbNBRX/view?usp=sharing)

**Comment by sarmientoj24** [comment]
I have 9 labels. The black part is labeled as 0 which is basically a bg class

**Comment by qubvel** [comment]
I got 0.43 IoU score with your data

```python
import torch
import numpy as np
import segmentation_models_pytorch as smp

target = np.load("masks.npy")
output = np.load("out.npy")

target = torch.tensor(target)
output = torch.tensor(output)

target = target.long()
output = output.argmax(1)

tp, fp, fn, tn = smp.metrics.get_stats(output, target, "multiclass", num_classes=9)
iou_score = smp.metrics.iou_score(tp, fp, fn, tn, "micro")
# 0.43
```

**Comment by sarmientoj24** [comment]
Hi. Didnt you use softmax on the output?

**Comment by sarmientoj24** [comment]
That's pretty weird. I am getting 1.
![image](https://user-images.githubusercontent.com/8830319/150526246-14416ec4-2642-443e-b4be-e14e1565dde2.png)

**Comment by qubvel** [comment]
> Hi. Didn't you use softmax on the output? Why is that?

No, it is not needed if you use argmax

**Comment by qubvel** [comment]
> That's pretty weird. I am getting 1. ![image](https://user-images.githubusercontent.com/8830319/150526246-14416ec4-2642-443e-b4be-e14e1565dde2.png)

Yes, that's strange... I don't know what the reason is and can just advise checking on other machines or in Colab, maybe reinstalling the package will help.

**Comment by sarmientoj24** [comment]
So I do not need to use softmax here for example?
```
                reconstructed_mask = out[i].detach()
                reconstructed_mask = torch.nn.functional.softmax(
                    reconstructed_mask.to(torch.float32), dim=1
                )
                reconstructed_mask = torch.argmax(reconstructed_mask, dim=0)
                reconstructed_mask = reconstructed_mask.cpu().numpy()
```

**Comment by sarmientoj24** [comment]
This is just weird:  
My machine
![image](https://user-images.githubusercontent.com/8830319/150532912-1befd6ab-1a9f-4d9b-ad72-0b0057a0633e.png)

Colab
![image](https://user-images.githubusercontent.com/8830319/150532956-57312eae-7877-4b9b-bdc5-8430c0138e6c.png)


Why is this a thing :(

**Comment by sarmientoj24** [comment]
My torch is `1.10.0` while Colab is `1.10.0+cu111`

**Comment by qubvel** [comment]
> So I do not need to use softmax here for example?
> 
> ```
>                 reconstructed_mask = out[i].detach()
>                 reconstructed_mask = torch.nn.functional.softmax(
>                     reconstructed_mask.to(torch.float32), dim=1
>                 )
>                 reconstructed_mask = torch.argmax(reconstructed_mask, dim=0)
>                 reconstructed_mask = reconstructed_mask.cpu().numpy()
> ```

1. If you dont use it for loss computation - softmax is not needed.
2. You are using softmax with dim=1 but argmax with dim=0, check it

**Comment by qubvel** [comment]
Could you compare where the difference is, in tp/fp/.. or in metric computation itself? Print out for example true positives for both cases

**Comment by qubvel** [comment]
@sarmientoj24 Is there any news? were you able to find the reason?

**Comment by sarmientoj24** [comment]
No. I still cannot figure it out. It works on all other machines but not mine. One thing i am suspicious is that it might have something to do with CUDA/cuDNN libraries since I have a very new graphics card

**Comment by davodogster** [comment]
Hi @qubvel  and @sarmientoj24 

I'm trying and struggling to train SMP with PY Lightning for multiclass (masks are single channel integer values (0, n_classes-1).

@sarmientoj24 I noticed some functions are missing in your code?? e.g add_plus_one and convert_binary_mask.

reconstructed_mask = add_plus_one_except(reconstructed_mask)
reconstructed_mask = convert_binary_mask(reconstructed_mask, colors=self.label_colors.

Were you able to successfully train multiclass with SMP + PTL?
@qubvel can you produce a tutorial notebook?
Thanks guys.

**Comment by AMMISAIDFaical** [comment]
i am in same problem getting 1. for all metrics

**Comment by ed18mtr** [comment]
Same here 1. for all metrics
